oci-image/src/state.rs

```diff
@@ -0,0 +1,212 @@
+pub struct LayerMetadata {
+    /// The id of the layer
+    pub id: String,
+    /// The compressed layer digest
+    pub compressed_digest: String,
+    /// The decompressed layer digest,
+    pub uncompressed_digest: String,
+    /// The path where the layer is stored
+    pub store_path: String,
+}
+
+/// `ImageMetadata` struct holds information's about an image in the state.
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct ImageMetadata {
+    /// The image id
+    pub id: String,
+    /// The image reference, such as `docker.io/amd64/ubuntu`
+    pub reference: String,
+    /// The image digest
+    pub digest: String,
+    /// The image layers metas
+    pub layers: Vec<LayerMetadata>,
+    /// The image configuration
+    pub config: ImageConfiguration,
+}
+
+/// `State` is responsible about storing information's about images and layers.
+#[derive(Clone, Default, Deserialize, Debug, Serialize, PartialEq)]
+pub(crate) struct State {
+    /// An hashmap that holds every images pulled
+    images: HashMap<String, ImageMetadata>,
+    /// An hashmap that holds every layers pulled
+    layers: HashMap<String, LayerMetadata>,
+    /// An hashmap that holds the current index for a specific snapshot type
+    snapshots: HashMap<String, usize>,
+}
+
+impl State {
+    /// Save the state into the file.
+    pub fn save(&self, path: &Path) -> crate::Result<()> {
+        let serialized = serde_json::to_string_pretty(&self)
+            .map_err(|e| Error::SerializeState(e.to_string()))?;
+
+        OpenOptions::new()
+            .write(true)
+            .open(path)
+            .map_err(|e| Error::OpenStateFile(e.to_string()))?
+            .write_all(serialized.as_bytes())
+            .map_err(|e| Error::WriteStateFile(e.to_string()))
+    }
+}
+
+impl MetadataManager for State {
+    fn add_image(&mut self, image: &ImageMetadata) -> &mut State {
+        self.images.insert(image.id.clone(), image.clone());
+        self
+    }
+
+    fn add_layer(&mut self, layer: &LayerMetadata) -> &mut State {
+        self.layers
+            .insert(layer.compressed_digest.clone(), layer.clone());
+        self
+    }
+
+    fn add_snapshot(&mut self, snapshot_type: &SnapshotType, index: usize) -> &mut Self {
+        self.snapshots.insert(snapshot_type.to_string(), index);
+        self
+    }
+
+    fn has_image(&self, image_id: &str) -> bool {
+        self.images.contains_key(image_id)
+    }
+
+    fn has_layer(&self, layer_id: &str) -> bool {
+        self.layers.contains_key(layer_id)
+    }
+
+    fn has_snapshot(&self, snapshot_id: &str) -> bool {
+        self.layers.contains_key(snapshot_id)
+    }
+
+    fn image(&self, image_id: &str) -> Option<&ImageMetadata> {
+        self.images.get(image_id)
+    }
+
+    fn layer(&self, layer_id: &str) -> Option<&LayerMetadata> {
+        self.layers.get(layer_id)
+    }
+
+    fn snapshot_index(&self, snapshot_type: &SnapshotType) -> usize {
```

Comment:
That's the only one you care about. The implementation is correct, but after increasing the index, you should store it back to the state file.